src/composition.rs

```diff
@@ -534,7 +559,9 @@ impl Composition {
         Self { file }
     }
 
-    pub(crate) fn composition_string(&self) -> Option<String> {
+    pub(crate) fn composition_string(
+        &self,
+    ) -> Option<(String, (ast::Position, ast::Position))> {
```

Comment:
I have two thoughts here: (a) a function called `composition_string` returning more than just a string seems awkward to me, and maybe ends up being a code smell? Could we get position information as separate call, without a performance hit? (b) Could `(ast::Position, ast::Position)` be better expressed as `ast::Range`?

Reply:
I think it's more about naming convention for this method. We originally only had the `to_string()` for composition, which returns the entire editor string (string generation on the entire File).

For the LSP applyEdit, we want to know the replacement position and replacement text. We use `composition_string()` both on the old_text, and the new_text, and then compare the differences. For the old_text, we only care about existence and position. For the new_text, we care about both the text itself and position.

So perhaps we should change the name to `get_composition_block()`?

Reply:
There is an `ast::SourceLocation` which is like the range. It also includes the File and source string. I could update the signature to return this structure.